### ISSUE TITLE
plone.rest breaks theme file uploads

### ISSUE THREAD

**Issue by displacedaussie** [plone/plone.rest#59]
# Problem

In the Plone theme editor it's possible to upload files.

Via the upload pattern, this ends up attempting to POST the file to something like: 

http://localhost:8080/Plone/portal_resources/theme/test-theme/themeFileUpload

However this request gets flagged as an `IAPIRequest` by `plone.rest` and is therefore handled by `RESTTraverse`:

https://github.com/plone/plone.rest/blob/master/src/plone/rest/traverse.py#L17

```
(Pdb) self
<plone.rest.traverse.RESTTraverse object at 0x113613810>
(Pdb) request
<HTTPRequest, URL=http://localhost:8080/Plone/portal_resources>
(Pdb) name
'portal_resources'
```

During traversal it gets the portal_resources object, but this isn't `IContentish` or an `IService` and so it raises a KeyError rather than returning the object.

```
(Pdb) obj
<BTreeFolder2 at /Plone/portal_resources>
(Pdb) IContentish.providedBy(obj)
False
(Pdb) IService.providedBy(obj)
False
```

https://github.com/plone/plone.rest/blob/master/src/plone/rest/traverse.py#L25

When handling the Exception, there are no valid view lookups so it raises the Error. This then results in a 404 to the themeFileUpload URL and a "File transfer failed" in the theme file upload UI.

# Options

 * Check for more than IContentish and IService during traversal so that objects like this are returned correctly?
 * Don't flag this request as an IAPIRequest so it's handled by normal traversal?
 * Configure something elsewhere (in `plone.resource`?) so that traversal works as expected?

**Comment by tisto** [comment]
@displacedaussie thank you for your detailed bugreport! Would you mind providing us with the full details of the POST request (e.g. headers, payload)? This makes it easier for us to understand why this request ends up in plone.rest. If you want to help us to fix this problem even quicker, you could write a failing unit/integration test with that particular request.

**Comment by tisto** [comment]
@displacedaussie can you please provide us with the full POST request? I still don't fully understand what goes on.

**Comment by djay** [comment]
@tisto  Here is sample request from the theme editor. I've also put a failing test in #72 

Haven't created a test yet but here is what the request looks like
```
General:
Request URL:http://localhost:8080/Plone/portal_resources/theme/asdsd/themeFileUpload?_authenticator=cdabc0f0a58dd5f5b910d36c2d91bd2452a94479
Request Method:POST
Status Code:200 OK
Remote Address:127.0.0.1:8080
Referrer Policy:no-referrer-when-downgrade

Response headers:
Content-Length:21
Content-Type:application/json
Date:Fri, 16 Feb 2018 08:11:20 GMT
Server:Zope/(2.13.26, python 2.7.14, darwin) ZServer/1.1
X-Frame-Options:SAMEORIGIN

Request headers:
Accept:application/json
Accept-Encoding:gzip, deflate, br
Accept-Language:en-US,en;q=0.9
Authorization:Basic YWRtaW46YWRtaW4=
Cache-Control:no-cache
Connection:keep-alive
Content-Length:2253194
Content-Type:multipart/form-data; boundary=----WebKitFormBoundarymOeiMFE1V2Acr015
Cookie:_pk_id.1.1fff=f9e2a914c3bc1362.1507226287.1.1507226468.1507226287.; __qca=P0-866781254-1511770439608; _ga=GA1.1.1561024462.1511770438
Host:localhost:8080
Origin:http://localhost:8080
Referer:http://localhost:8080/Plone/++theme++asdsd/@@theming-controlpanel-mapper
User-Agent:Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.167 Safari/537.36
X-Requested-With:XMLHttpRequest

Query string paramaters:
_authenticator:cdabc0f0a58dd5f5b910d36c2d91bd2452a94479

Request payload:
------WebKitFormBoundarymOeiMFE1V2Acr015
Content-Disposition: form-data; name="file"; filename="Screen Shot 2018-02-16 at 3.08.15 pm.png"
Content-Type: image/png


------WebKitFormBoundarymOeiMFE1V2Acr015--
```

**Comment by djay** [comment]
@tisto possibly there should also be a test in https://github.com/plone/plone.app.theming/blob/master/src/plone/app/theming/tests/test_controlpanel.py. Seems like there are no tests for the theme editor end points. 
But that would only pickup an error during 5.1 when plone.rest is merged. Seems like plone.rest should test it doesn't get in the way of stuff thats already in plone.

**Comment by djay** [comment]
@tisto the code that is supposed to be called is https://github.com/plone/plone.app.theming/blob/master/src/plone/app/theming/browser/themefile.py#L25

**Comment by tisto** [comment]
@djay if that code sets the Accept header to application/json it expects a JSON structure as a response. I'd check first if this is the case and then if it is necessary. I don't see why standard Plone code would need to do something like this. plone.rest is the wrong place to fix this IMHO.

**Comment by djay** [comment]
@tisto plone code does expect and get a JSON code in response. The problem is that plone.rest is blocking the normal plone code from giving that response. Assuming that plone.rest is the only thing in plone that uses rest is a failure in plone.rest. plone.rest already has exceptions to where it lets the normal plone code return a json response. This is adding one more. What is the problem with that?

**Comment by djay** [comment]
@tisto dropzone is thirdparty code. it's not made by plone
Also this is not just theming. All the code that uses dropzone will be affected I believe. File widgets, tinymce, folder contents. I'm not sure why plone.rest is not interfering with them currently however but presumably the plone code has to give the same JSON response. 
Another way to think of it is this. If plone.rest is not doing anything useful in portal_resources then why is it interfering with responses from it?

**Comment by NicolasGoeddel** [comment]
Anything new here or does someone have a good workaround until this gets fixed? I would really appreciate your help here.

**Comment by tisto** [comment]
@NicolasGoeddel no. This is still an issue that needs to be fixed in plone.app.theming. This is not a plone.rest issue.

**Comment by NicolasGoeddel** [comment]
@tisto Thank you. I did not get that. It looks like there is no issue about that in plone.app.theming.

Can someone tell me what I need to change in themefile.py to make it work? Do I have to return a different HTTP response instead of the JSON value?

**Comment by tisto** [comment]
@NicolasGoeddel sure. Please see my previous comments:

https://github.com/plone/plone.rest/issues/59#issuecomment-422822812
https://github.com/plone/plone.rest/issues/59#issuecomment-422824149

The problem is that any request that uses the "Content-Type" header "application/json" hooks into Plone REST API. We have to amend plone.app.theming to return a status code that complies with the HTTP standard (return code 200 or 422) instead of sending the status within the body of the response (which is an anti-pattern). Then we should remove setting "application/json" as HTTP header in the response and amend the frontend code accordingly.

I am happy to help with the REST API stuff. Though, I never touch patternslib/mockup in my dev life...

**Comment by djay** [comment]
@tisto Your suggestion is to change the api in this specific case to not expect JSON (ie change the expect in the request not response as you seem to suggest). The api could be changed in this particular case to not use json but the problem could come up with other parts of Plone in the future.
Your suggestion therefore seems to be to not use REST or any api that requires a json response in any part of plone except for plone.rest. plone.rest does not have an api that applies to portal resources. If it did then the upload code could be modified to use that. You are just kicking the issue down the road.
A better solution is to ensure that plone.rest does not interfere with apis in other parts of plone that it does not apply.
@esteele  The fact that we had a patch for this issue years ago, created a blocking breaking test for this issue that should have prevented plone from being released... and it still was... seems like a break down of process to me.

**Comment by tisto** [comment]
@djay I guess we can agree that we disagree here.

**Comment by tisto** [comment]
Just for the record (again). The current plone.app.theming code is just doing it wrong. If we amend it to follow the HTTP spec and REST API best practices the problem is gone:

> The problem is that any request that uses the "Content-Type" header "application/json" hooks into Plone REST API. We have to amend plone.app.theming to return a status code that complies with the HTTP standard (return code 200 or 422) instead of sending the status within the body of the response (which is an anti-pattern). Then we should remove setting "application/json" as HTTP header in the response and amend the frontend code accordingly.

Amending the plone.rest code to work around badly written code is the wrong way to solve this problem IMHO.

**Comment by djay** [comment]
@tisto Just for the record (again). No one is disagreeing the pattern in plone.app.theming image upload code is bad. But that isn't causing the bug in this case. If it did use REST correctly it would still not work as plone.restapi would still block it from working. 
plone.restapi should not override all plone views that expect a json response unless it explicitly has a REST api endpoint for that part of Plone.

**Comment by ewohnlich** [comment]
We ran into this in 5.2 because plone.restapi is now in the Plone release by default. I understand there is a disagreement on whether it is plone.rest or plone.app.theming that needs an update (or both) but surely an arbiter should get involved if both those packages are going to be part of core now.

**Comment by tisto** [comment]
I will close this issue here. I made my point clear and discussed this with both the Plone FWT and the Plone release manager.

**Comment by sm2x** [comment]
> We ran into this in 5.2 because plone.restapi is now in the Plone release by default. I understand there is a disagreement on whether it is plone.rest or plone.app.theming that needs an update (or both) but surely an arbiter should get involved if both those packages are going to be part of core now.

so what is the workaround at the moment, we need the upload function?

**Comment by yurj** [comment]
The workaround is to develop the theme on fs. This bug is about people unable to agree on a fix. @djay : someone should add a test in plone to show upload fails, so tests does not pass and they can't release a new version without fixing it.

**Comment by djay** [comment]
@yurj I believe I did that a couple of years ago.

**Comment by djay** [comment]
@yurj I'm not sure why my test failed to fail. But have a go changing it to make it fail - https://github.com/plone/plone.app.theming/pull/140

**Comment by djay** [comment]
@yurj I realise what I did wrong. the tests didn't install restapi. I've fixed this in https://github.com/plone/plone.app.theming/pull/195

**Comment by yurj** [comment]
> > We ran into this in 5.2 because plone.restapi is now in the Plone release by default. I understand there is a disagreement on whether it is plone.rest or plone.app.theming that needs an update (or both) but surely an arbiter should get involved if both those packages are going to be part of core now.
> 
> so what is the workaround at the moment, we need the upload function?

you can also use zmi -> portal_resources -> theme and upload a File there.

**Comment by petri** [comment]
Note: This would be fixed if/when #105 gets merged.
